File: go/token/token.go
package token

type TokenType string

type Token struct {
	Type    TokenType
	Literal string
}

const (
	ILLEGAL = "ILLEGAL"
	EOF     = "EOF"

	// Identifiers + literals
	IDENT  = "IDENT"
	INT    = "INT"
	FLOAT  = "FLOAT"
	STRING = "STRING"

	// Operators
	ASSIGN   = "="
	PLUS     = "+"
	MINUS    = "-"
	ASTERISK = "*"
	SLASH    = "/"

	EQ     = "=="
	NOT_EQ = "!="
	LT     = "<"
	LTE    = "<="
	GT     = ">"
	GTE    = ">="

	// Delimiters
	COMMA     = ","
	SEMICOLON = ";"
	LPAREN    = "("
	RPAREN    = ")"
	LBRACE    = "{"
	RBRACE    = "}"

	// Keywords
	LET   = "LET"
	CONST = "CONST"
	PRINT = "PRINT"
	IF    = "IF"
	ELSE  = "ELSE"
	WHILE = "WHILE"
	FOR   = "FOR"
	TRUE  = "TRUE"
	FALSE = "FALSE"

	FUNC   = "FUNC"
	CLASS  = "CLASS"
	RETURN = "RETURN"
)

var keywords = map[string]TokenType{
	"let":    LET,
	"const":  CONST,
	"print":  PRINT,
	"if":     IF,
	"else":   ELSE,
	"while":  WHILE,
	"for":    FOR,
	"true":   TRUE,
	"false":  FALSE,
	"func":   FUNC,
	"class":  CLASS,
	"return": RETURN,
}

func LookupIdent(ident string) TokenType {
	if tok, ok := keywords[ident]; ok {
		return tok
	}
	return IDENT
}

File: rust/src/token/token.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenType {
    Illegal,
    Eof,

    // Identifiers + literals
    Ident,
    Int,
    Float,
    String,

    // Operators
    Assign,
    Plus,
    Minus,
    Asterisk,
    Slash,
    Eq,
    NotEq,
    Lt,
    Lte,
    Gt,
    Gte,

    // Delimiters
    Comma,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,

    // Keywords
    Let,
    Print,
    If,
    Else,
    While,
    True,
    False,
    Func,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
}

impl Token {
    pub fn new(token_type: TokenType, literal: &str) -> Self {
        Self {
            token_type,
            literal: literal.to_string(),
        }
    }
}

impl TokenType {
    pub fn lookup_ident(ident: &str) -> TokenType {
        match ident {
            "let" => TokenType::Let,
            "print" => TokenType::Print,
            "if" => TokenType::If,
            "else" => TokenType::Else,
            "while" => TokenType::While,
            "true" => TokenType::True,
            "false" => TokenType::False,
            "func" => TokenType::Func,
            _ => TokenType::Ident,
        }
    }
}
